File: MathAndPrimeNumbers/SmallestValueAfterReplacingWithSumOfPrimeFactors/smallestValue.cpp
#include <cmath>
using namespace std;

class Solution {
public:
    bool isPrime(int n){
        if (n<=1) return false;

        for(int i=2;i<=int(sqrt(n));i++){
            if(n%i == 0) return false;
        }

        return true;
    }

    int smallestValue(int n) {
        if(isPrime(n))return n;
        int sum = 0;
        int m = n;
        for(int i=2;i<=sqrt(n);i++){
            while(m%i == 0 && isPrime(i)){
                sum += i;
                m /= i;
            }
        }
        for(int i=sqrt(n);i>0;i--){
            while(m%(n/i) == 0 && isPrime(n/i)){
                sum += (n/i);
                m /= (n/i);
            }
        }
        if(sum == n)return sum;
        return smallestValue(sum);
    }
};

File: MathAndPrimeNumbers/SmallestValueAfterReplacingWithSumOfPrimeFactors/smallestValue.py
from math import sqrt

class Solution:
    def isPrime(self, n):
        if n<=1: return False

        for i in range(2,int(sqrt(n)+1)):
            if n%i == 0: return False

        return True

    def smallestValue(self, n: int) -> int:
        if self.isPrime(n): return n
        sum = 0
        m = n
        for i in range(2,int(sqrt(n)+1)):
            if m%i == 0 and self.isPrime(i):
                while m%i == 0: 
                    sum += i
                    m //= i
        
        for i in range(int(sqrt(n)),0,-1):
            if m%(n//i) ==0 and self.isPrime(n//i):
                while m % (n//i) == 0: 
                    sum += (n//i)
                    m //= (n//i)

        if sum == n: return n
        
        return self.smallestValue(sum)
        